server/blob/blob-cloud/src/main/java/org/apache/james/blob/cloud/CloudsBlobsConfiguration.java

```diff
@@ -0,0 +1,115 @@
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.cloud;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.Properties;
+
+import org.jclouds.openstack.keystone.config.KeystoneProperties;
+import org.jclouds.openstack.swift.v1.reference.TempAuthHeaders;
+
+public class CloudsBlobsConfiguration {
+    private final URI endpoint;
+    private final Optional<Region> region;
+    private final Identity identity;
+    private final Properties overrides;
+    private final Credentials credentials;
+
+    public static class Builder {
+        private URI endpoint;
+        private Identity identity;
+        private Credentials credentials;
+        private Optional<Region> region;
+        private final Properties overrides;
+
+        public Builder() {
+            region = Optional.empty();
+            this.overrides = new Properties();
+        }
+
+        public CloudsBlobsConfiguration.Builder withEndpoint(URI endpoint) {
+            this.endpoint = endpoint;
+            return this;
+        }
+
+        public CloudsBlobsConfiguration.Builder withIdentity(Identity identity) {
+            this.identity = identity;
+            return this;
+        }
+
+        public CloudsBlobsConfiguration.Builder withCredentials(Credentials credentials) {
+            this.credentials = credentials;
+            return this;
+        }
+
+        public CloudsBlobsConfiguration.Builder withRegion(Region region) {
+            this.region = Optional.of(region);
+            return this;
+        }
+
+        public CloudsBlobsConfiguration.Builder withTempAuthHeaderUserName(String tmpAuthHeaderUser) {
+            this.overrides.setProperty(TempAuthHeaders.TEMP_AUTH_HEADER_USER, tmpAuthHeaderUser);
+            return this;
+        }
+
+        public CloudsBlobsConfiguration.Builder withTempAuthHeaderPassName(String tmpAuthHeaderUser) {
+            this.overrides.setProperty(TempAuthHeaders.TEMP_AUTH_HEADER_PASS, tmpAuthHeaderUser);
+            return this;
+        }
+
+        public CloudsBlobsConfiguration build() {
+            overrides.setProperty(KeystoneProperties.CREDENTIAL_TYPE, "tempAuthCredentials");
```

Comment:
That's a good question, I wasn't too sure how to go about it. Checking in the build method means a runtime error. 
Ideally I would prefer the builder constructor to force the mandatory parameters and have fluent methods for the optional ones would that be acceptable ?

Reply:
usally we are checking in the build method. Because checking (and failing) early is better than failing too late.
I you want to do a progressive builder enforcing mandatory parameters you are welcome of course (if you are not already bored with boilerplate)